Tech Preview feature not working for Integration Editor

kahboom opened the pull request:
Created another version of `ConnectionsWithToolbar` per @riccardo-forina 's suggestion.

<img width="1330" alt="Screenshot 2019-06-04 16 06 10" src="https://user-images.githubusercontent.com/3844502/58890595-d7a29c80-86e2-11e9-8b8c-9fa96969b602.png">

fix #374

kahboom committed:
Fix duplicate connection prop preventing tech preview check from working in editor

Assign string to var instead of using string literal

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2146/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

gashcrumb commented on app/ui-react/syndesis/src/modules/connections/components/ConnectionsWithToolbarNew.tsx:
```diff
@@ -0,0 +1,126 @@
+  children?: any;
+}
+
+export class ConnectionsWithToolbarNew extends React.Component<
```
U need a better name.  I wonder, maybe this should live in the `integrations` module and maybe be called like I dunno, `AllStepsWithToolbar` or something?  That way it's a bit more obvious that this guy isn't just for "connections" in the syndesis sense.  Ooh ooh, maybe `UIStepsWithToolbar`

gashcrumb replied:
Works great BTW!

kahboom replied:
Yeah I was struggling a bit with the name, as you can see. 😂 I think `UIStepsWithToolbar` sounds good. 👍

gashcrumb reviewed (changes requested):
The name, it has to change :smile:

kahboom committed:
Rename to UIStepsWithToolbar and move to integrations module

riccardo-forina commented on app/ui-react/syndesis/src/modules/connections/components/ConnectionsWithToolbarNew.tsx:
```diff
@@ -0,0 +1,126 @@
+import { Connections, IConnectionsProps } from './Connections';
+
+function getFilteredAndSortedConnections(
+  connections: Connection[],
```
we are not dealing with `Connection` objects here, but with `IUIIntegrationStep` which is a "meta object" that can hold `Connection`s, `StepKind`s and `Extension`s

riccardo-forina commented on app/ui-react/syndesis/src/modules/connections/components/ConnectionsWithToolbarNew.tsx:
```diff
@@ -0,0 +1,126 @@
+                  })}
+                >
+                  {this.props.children}
+                  <Connections
```
Because of the above, we can't reuse the `Connections` component.

riccardo-forina reviewed (changes requested):
I suggest `EditorStep` for the name.

kahboom committed:
create EditorSteps and EditorStepsWithToolbar for reuse in integrations module with IUIIntegrationStep type

kahboom committed:
Rework the typings for compatibility

Lint fixes

kahboom commented:
Alright, here are the changes I've made, based on feedback, that seem to work (albeit some are hacky):
- Rename to `EditorStepsWithToolbar`
- Move to `integrations` module for reuse
- Duplicate of `Connections` component called `EditorSteps` that uses the `IUIIntegrationStep` interface and `EditorStepsWithToolbar`
- Adjusted the typings to comply (please note the use of `as Blah[]` a few times, not ideal but it works for now)
- Rename `connections` to `steps` to reflect data more accurately

And still using the `.map()` part for the `filteredAndSortedConnections`, as we needed it.

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorSteps.tsx:
```diff
@@ -0,0 +1,166 @@
+                                      'configurationRequired'
+                                    )}
+                                    i18nTechPreview={t('techPreview')}
+                                    menuProps={
```
I don’t think we have to show a menu at all here

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorStepsWithToolbar.tsx:
```diff
@@ -0,0 +1,126 @@
+  activeFilters.forEach((filter: IActiveFilter) => {
+    const valueToLower = filter.value.toLowerCase();
+    filteredAndSortedConnections = filteredAndSortedConnections.filter(
+      (c: Connection) => c.name.toLowerCase().includes(valueToLower)
```
This cast doesn’t feel right. Name should be a common property

gashcrumb replied:
Yeah, I'd suggest continuing to rename these variables for this file, in this case it's `filteredAndSortedEditorSteps` not connections.  and then not `c`, maybe `editorStep` or `es`.

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorStepsWithToolbar.tsx:
```diff
@@ -0,0 +1,126 @@
+                helpers.currentSortType,
+                helpers.isSortAscending
+              ).map(connection => {
+                return connection[connectionString];
```
Not all items have a connection property I think? I’ll have to check this tomorrow against a BE but I think we might be ending up not showing all the steps

gashcrumb replied:
Yeah, only `endpoint` kinds would have a `connection` property.  But we could be misled here if  it's just because `connection` needs to be renamed to something like `editorStep`.

kahboom replied:
Renaming to `editorStep[connectionString]`. I can do a check for `uiStepKind === 'editor-step'`, but at the moment only one of all the listed items isn't `endpoint`, and that's `api-provider`. Also, from what I'm seeing, even `api-provider` has a `connection` property. This is how the list appears for me prior to `.map()`: https://gist.github.com/kahboom/d58532286c0f94da46ef17e0c477ba73

In the raw version, grep `api-provider` as an example.

riccardo-forina replied:
`editor-step` [isn't a valid value](https://github.com/syndesisio/syndesis-react/blob/4cbbdfbe57f0e08ea133138551a7752dccacf737/app/ui-react/syndesis/src/modules/integrations/components/editor/interfaces.ts#L34) for uiStepKind.

Again, we are not dealing with connections, but with a mixed bag of different things: connections, stepkinds and extensions. All those types [have been normalized as `IUIStep` objects](https://github.com/syndesisio/syndesis-react/blob/e3b8fc2d0ae9454ce7edf633f5f11afce6994abf/app/ui-react/syndesis/src/modules/integrations/components/editor/SelectConnectionPage.tsx#L102-L106). You'll have a valid `title` - to be used for the sorting an filtering - and for some objects a valid `connector` object. All will have some `metadata` set.
All this info needs to be properly handled to show - or not - the tech preview/configuration required badges.

kahboom replied:
I'm aware that we are not dealing with connections, but what I'm getting for the data is an array of objects each with a parent `connection` property.
```
[
  { connection: { configuredProperties: { .. }, connector: { .. }, metadata: { .. }  } },
  { connection: { .. } },
  { connection: { .. } },
  and so on..
]
```
I can either `.map()` it or similar, or I can write some type of interface like
```
export interface IUIEditorStep extends IUIIntegrationStep {
  connection: IUIStep
}
```
Neither is ideal, but if you have a different approach in mind please share what you are looking for exactly.

The original purpose of the PR was to fix the Tech Preview feature that was previously working, if we want to roll out a new toolbar for connections and are changing the object, that's getting kinda out of the scope of this PR.

gashcrumb replied:
Can you re-read what @riccardo-forina said?  You can't rely on `connection` being set for this particular collection.  Much like the old angular app tried to re-use the "Connection List" page we actually kinda started in the same path here in the React app, as I recall when this page was originally written I had just created the combined step/connection/extension step in the angular app.  So this component that you're writing has to deal with that same combination.  At the very least, these objects are normalized into an IUIStep, so you should only need to look at the `metadata` value for each thing in the collection, not the `connection` object directly.  Also, use the `getMetadataValue<string>`, helper just in the off chance an item in the collection doesn't have it set.  I haven't taken a close look to see if you're using that here.

riccardo-forina reviewed (commented):
Just a quick check, I’ll test this more thoroughly tomorrow

gashcrumb commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorSteps.tsx:
```diff
@@ -0,0 +1,166 @@
+            {({ pushNotification }) => {
+              return (
+                <WithConnectionHelpers>
+                  {({ deleteConnection }) => {
```
Delete connection isn't needed here at all, this page doesn't require any management functions.

gashcrumb replied:
That means you can get rid of the notifications, `WithConnectionHelpers` etc

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2164/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

kahboom committed:
remove deleteConnections, menu, notifications

kahboom committed:
rename connections to editorSteps

Re-add map for flag to work properly

Lint fix

Update the icon

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2188/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2201/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2222/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

riccardo-forina reviewed (changes requested):
It doesn't work

![doesntwork](https://user-images.githubusercontent.com/966316/59016783-56621b80-8842-11e9-9087-8775e78a6638.gif)

riccardo-forina commented:
Actually, this needs a rebase, the error I posted above is because we have an extension on staging and this PR is based on a version of the app that was bugging out with that.

kahboom committed:
Use getMetadataValue to check for metadata, rm map

kahboom commented:
<img width="1673" alt="Screenshot 2019-06-06 17 20 10" src="https://user-images.githubusercontent.com/3844502/59049588-3bae9780-8880-11e9-8066-8b68f6a01ca6.png">

No additional or changes to interfaces, no `.map`, no referencing the `Connection` interface from what I can see, just added an additional check and used @gashcrumb 's suggestion of the `getMetadataValue` helper. I did read @riccardo-forina 's suggestion several times. I do understand that this is a not a Connection and instead comprises connections, extensions, and stepKinds, where `IUIStep` normalizes it. This is using that interface. Something happens in `mergeConnectionsSources` where a `connection` property gets added to each object. I understand we can't rely on that, but it's still there, and needs to be iterated over to reach `metadata`, so I've added another check in `EditorSteps.tsx` to check if that property exists. Short of that I'm out of ideas, so let me know what exactly would need to change, if anything.

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2240/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

riccardo-forina commented:
Did you test something different than the first step flow? Try adding a 3rd step, maybe in one of the existing integrations, the collection will be a mixed thing. In the first step, only connections and extensions are allowed

kahboom committed:
Check for connection prop

kahboom commented:
Good catch, I should have checked for `connection`, let me know if this suffices.

<img width="1677" alt="Screenshot 2019-06-06 19 31 45" src="https://user-images.githubusercontent.com/3844502/59057218-ccda3a00-8891-11e9-85dd-9da2c82c6be5.png">

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2244/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorStepsWithToolbar.tsx:
```diff
@@ -0,0 +1,119 @@
+                    loading={this.props.loading}
+                    steps={filteredAndSortedEditorSteps}
+                    getConnectionHref={this.props.getConnectionHref}
+                    getConnectionEditHref={this.props.getConnectionEditHref}
```
This isn't needed

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorSteps.tsx:
```diff
@@ -0,0 +1,99 @@
+  loading: boolean;
+  steps: IUIStep[];
+
+  getConnectionHref(connection: ConnectionOverview): H.LocationDescriptor;
```
we are not dealing with just connections, so better not use `connection` when naming things. Also, the object that's passed back should really be a `IUIStep`, shouldn't it?

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorSteps.tsx:
```diff
@@ -0,0 +1,99 @@
+                        'tech-preview',
+                        s!.connection!.connector!.metadata
+                      ) === 'true';
+                  } else if (s.connector!) {
```
the `!` isn't needed here

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorSteps.tsx:
```diff
@@ -0,0 +1,99 @@
+                    isTechPreview =
+                      getMetadataValue<string>(
+                        'tech-preview',
+                        s.connector!.metadata
```
same here, the above makes the object valid

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorSteps.tsx:
```diff
@@ -0,0 +1,99 @@
+                    isTechPreview =
+                      getMetadataValue<string>(
+                        'tech-preview',
+                        s!.connection!.connector!.metadata
```
drop the `!`, the above `if` is making those superfluous

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorStepsWithToolbar.tsx:
```diff
@@ -0,0 +1,119 @@
+  activeFilters.forEach((filter: IActiveFilter) => {
+    const valueToLower = filter.value.toLowerCase();
+    filteredAndSortedEditorSteps = filteredAndSortedEditorSteps.filter(
+      (e: Connection) => e.name.toLowerCase().includes(valueToLower)
```
Again, we are not dealing with just connections...

kahboom committed:
Update name for props, interface

Rm getConnectionHref

kahboom commented:
Okay, I think it's fine now, but anything else, let me know. The specific feedback really helped both in terms of time and getting a better understanding of what I was missing.

<img width="1678" alt="Screenshot 2019-06-07 10 35 56" src="https://user-images.githubusercontent.com/3844502/59095452-6ba57b80-8910-11e9-8e7a-04e781d1d451.png">

riccardo-forina commented:
It doesn't build

kahboom commented:
Sorry I forgot to commit the last change.

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2269/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

gashcrumb commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorSteps.tsx:
```diff
@@ -0,0 +1,97 @@
+            >
+              {() => {
+                return this.props.steps.map((s, index) => {
+                  const configurationRequired =
```
I feel like this could be refactored out into a helper function, probably a function that could take a `Connection | IUIStep`, like:

```
isConfigurationRequired(thingWithBoard: Connection | IUIStep)
```

That way we could have one block of code in the codebase that makes this determination.  I haven't checked but I imagine we could have this pattern in a couple places at this point.

gashcrumb replied:
Other option is to have the function just accept the bulletin board, actually I like that better, i.e.:

```
isConfigurationRequired(board: BulletinBoard)
```

Then you in all cases you could just call `configurationRequired = isConfigurationRequired(s.board);` and be done with it :-)

gashcrumb replied:
BTW I wouldn't apply this refactoring to the integration list if you go this route, that flag shows up under slightly different circumstances.

kahboom replied:
Just did this but need to go back and refactor the other parts of the app that could use this, I can do it today but I think it should be separate from this PR.

gashcrumb replied:
I think it'd be worthwhile to do so now to be honest while you're in there anyways, no?

gashcrumb commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorSteps.tsx:
```diff
@@ -0,0 +1,97 @@
+
+                  let isTechPreview = false;
+
+                  if (s.connection && s.connection!.connector) {
```
Similarly with this block.  I'd say it'd be more worthwhile in this case to accept an `IUIStep`, `Connection` and `Connector` object.  Ideally it'd be cool if this could be reduced to:

```
const isTechPreview = figureOutTechPreviewFlag(s);
```

And then of course sort the other spots in the codebase where we're figuring this out.

kahboom replied:
This would be good, but I think should be addressed in a separate PR.

gashcrumb replied:
Why?  You're in the code now, it's probably the best time to tackle this while it's fresh, ya know?  It'd be a matter of writing functions that just return a boolean, then this block here and [this block](https://github.com/syndesisio/syndesis-react/blob/master/app/ui-react/syndesis/src/modules/connections/components/Connections.tsx#L79-L91) will do the same thing then.  Seems like a win to me 'cause you'll be reducing code.

gashcrumb commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorStepsWithToolbar.tsx:
```diff
@@ -0,0 +1,117 @@
+  IEditorStepsWithToolbarProps
+> {
+  public static defaultProps = {
+    includeHidden: false,
```
I think we want to show the hidden things in this particular case, no?  AFAIK this controls if the API provider connection is displayed and Timer and WebHook.  Actually, since this is now a component dedicated to just this page, I don't think we even need the flag.

gashcrumb replied:
Yeah, the page works as I'd expect, I think this might be a leftover.

kahboom replied:
Yeah this is just left over. Wasn't really looking at it from this perspective tbh, but it makes sense, I'll remove it.

kahboom committed:
Refactor check for configurationRequired

kahboom committed:
Rm unnecessary default values obj

Lint fixes

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2361/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

kahboom committed:
Rm unused Connection model

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2366/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

kahboom committed:
Create reusable check for tech preview flag

kahboom committed:
Refactor Connections.tsx to use helper functions

kahboom commented:
Changes based on feedback:
- Added `isConfigurationRequired` helper function in `connectionFunctions.ts`. When Riccardo gets [this](https://github.com/syndesisio/syndesis-react/pull/444/files#diff-c8c891bfe81f60cb81839bf4031cb8c7R5) in, we can probably move it there.
- Added `figureOutTechPreviewFlag` helper function in `modules/integrations/editor/utils.ts`. The reason for this, despite that the tech preview flag is used in other places, is that we cannot reference `IUIStep` from the `connectionFunctions.ts` API helpers file, which is defined in the `modules/integrations/editor/interfaces.ts` file.
- Refactored and referenced those functions where possible, which at the moment is limited to `Connections.tsx` and `EditorSteps.tsx`. Didn't do this for `Integrations.tsx` for the reasons mentioned above about the special conditions.

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2391/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

gashcrumb commented on app/ui-react/syndesis/src/modules/integrations/components/editor/utils.ts:
```diff
@@ -42,6 +44,44 @@ type StepKindHrefCallback = (
+        (data as IUIStep).connection!.connector!.metadata
+      ) === 'true'
+    );
+  } else if ((data as IUIStep).connector) {
```
So, you don't probably need this `if` statement, or the next one, at runtime this type info is meaningless 'cause it's gone.

kahboom replied:
If I do this instead, won't it always return `false` then for whenever the object is a `Connection` or is that a really dumb question? 😖 
```
export function figureOutTechPreviewFlag<
  T extends IUIStep | Connection | Connector
>(data: T): boolean {
  if ((data as IUIStep).connection && (data as IUIStep).connection!.connector) {
    return (
      getMetadataValue<string>(
        'tech-preview',
        (data as IUIStep).connection!.connector!.metadata
      ) === 'true'
    );
  } else if (data as Connector) {
    return getMetadataValue<string>('tech-preview', data.metadata) === 'true';
  } else {
    return false;
  }
}
```

kahboom replied:
I guess my question would be, then why would we want to accept `Connection` as a type? Would it make more sense to only accept `IUIStep` and `Connector`?

kahboom replied:
Hmm I suppose in places that it affects, like the Dashboard Connections list, I can pass a connector instead of connection, but the last question still applies I think. When I remove these `if` statements then the Dashboard Connections no longer displays the flag.

gashcrumb commented on app/ui-react/syndesis/src/modules/integrations/components/editor/utils.ts:
```diff
@@ -42,6 +44,44 @@ type StepKindHrefCallback = (
+ * returns a string value of 'true'
+ * @param data
+ */
+export function figureOutTechPreviewFlag<
```
Probably `isTechPreview` would be better, I know I said `figureOutTechPreviewFlag` but that was being a bit tongue-in-cheek, sorry!.

kahboom replied:
haha no worries, I thought so, but either way we won't be able to use the same declaration so maybe `isTechPreviewCheck` might work better? Or something similar?

kahboom committed:
Rename check for techPreview

Re-add checks until further clarification

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2395/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

kahboom committed:
Rm if statements, add configRequired prop to useConnection

kahboom commented:
Changes:

- Renamed the tech preview helper function to `isTechPreview`

- Removed the `if` statements as requested, although naturally this means that the Dashboard Connections flag for Tech Preview isn't working as expected (the `if` statements contained the checks that handled these). Please advise.

<img width="1669" alt="Screenshot 2019-06-11 09 30 27" src="https://user-images.githubusercontent.com/3844502/59256294-8edf6c00-8c2b-11e9-85aa-9a9152b8f0f6.png">

_Tech Preview flags no longer displaying for connections on Dashboard_

- I've added a `configRequired` property to `useConnection.tsx`, but until I fully understand what/why I'm doing with respect to the new changes, I don't think it's appropriate to refactor. If this is important then an issue should be raised, as it's not directly related to the bug that this PR is intending to resolve.

kahboom committed:
Pass connector directly to isTechPreview

kahboom commented:
In the meantime I've just decided to pass the `connector` from the Connections page (also used for the Dashboard Connections), so that it works again.

<img width="1657" alt="Screenshot 2019-06-11 09 47 24" src="https://user-images.githubusercontent.com/3844502/59257478-ec74b800-8c2d-11e9-8b9d-bafcc735efa8.png">

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2411/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

riccardo-forina commented on app/ui-react/syndesis/src/modules/connections/components/Connections.tsx:
```diff
@@ -78,16 +79,11 @@ export class Connections extends React.Component<IConnectionsProps> {
                               };
 
                               const configurationRequired =
```
Because of [this](https://github.com/syndesisio/syndesis-react/pull/409/files#diff-c8c891bfe81f60cb81839bf4031cb8c7R19) you don't need this anymore, `c.configRequired` should hold your value.

kahboom replied:
>I agree with Stan, we either do a full refactoring to smooth all the raw edges, or not.
>I'm ok with continuing working on this 'til everything is DRYed up and makes sense.

As long as you're ok with questions and helping to explain, I'm always happy to refactor.

`configRequired` is not a property of `ConnectionOverview`, will run into this for `isTechPreview` as well, I'm guessing--what is the best approach here? Should I make an interface that extends `ConnectionOverview`?

riccardo-forina replied:
I think we should be fine without a new interface. When you get connections out from `useConnections` (or `WithConnections`) the object that's returned is an object that does respect the `Connection` interface, but it's also extended with a number of properties that the app needs.

These connection objects are then eventually passed to the `toUIStep` function to convert them, and all their properties are spread in the new `IUIStep` one.
So if we simply extend the `IUIStep` interface to have `configRequired` and `isTechPreview` defined, we should be good.

kahboom replied:
I'm able to do this for `EditorSteps.tsx`, for instance, but not for `Connections.tsx` since it seems to never use `IUIStep`, where in `IConnectionsProps`: `connections: ConnectionOverview[];`. Should that instead be `IUIStep` or have I missed something?

riccardo-forina replied:
`Connetions.tsx` will use connections coming directly from the API, so already transformed. `c.configRequired` will be available then.

riccardo-forina replied:
But I see now that we might need to create an extended `IConnection` interface, since `Connections.tsx`. Boring, but yeah, needed.

riccardo-forina commented on app/ui-react/syndesis/src/modules/connections/components/Connections.tsx:
```diff
@@ -78,16 +79,11 @@ export class Connections extends React.Component<IConnectionsProps> {
-                                  ? c.connector!.metadata!['tech-preview'] ===
-                                    'true'
-                                  : false;
+                              const techPreview = c.connector
```
Can't this be treated like [this](https://github.com/syndesisio/syndesis-react/pull/409/files#diff-c8c891bfe81f60cb81839bf4031cb8c7R19)?

kahboom replied:
`techPreview` relied on `IUIStep` based on the feedback from before, which couldn't be access from those API helper files since it was within `modules/integrations/components/editor/interfaces.ts`. If we only accept `Connector` then it should be fine and I can get rid of the generics stuff as well.

riccardo-forina commented on app/ui-react/packages/api/src/helpers/connectionFunctions.ts:
```diff
@@ -82,6 +83,16 @@ export function getActionStepDefinition(
+export function isConfigurationRequired(
+  board: ConnectionBulletinBoard
+): boolean {
+  return (board!.notices || board!.warnings || board!.errors)! > 0;
```
`!`s are useless here, you defined board as a mandatory parameter.

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorSteps.tsx:
```diff
@@ -0,0 +1,81 @@
+              {() => {
+                return this.props.steps.map((s, index) => {
+                  const configurationRequired =
+                    s.board && isConfigurationRequired(s.board);
```
I think this is going to throw for non-connections, since `board` will be `undefined` and `isConfigurationRequired` is assuming board _will_ be defined.

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/utils.ts:
```diff
@@ -43,6 +45,30 @@ type StepKindHrefCallback = (
+ * returns a string value of 'true'
+ * @param data
+ */
+export function isTechPreview<T extends IUIStep | Connection | Connector>(
```
Do we really need to accept all these types? `Step`s don't have a `connection`, and neither `Extension`s do.

We really just want to accept `Connector` here, since the tech preview metadata is stored there.

Then [this](https://github.com/syndesisio/syndesis-react/pull/409/files#diff-c8c891bfe81f60cb81839bf4031cb8c7R10) can consume this function and set the `isTechPreview` flag.

Bonus points for:
* enhancing the `useConnector` to provide a transformation function (like the `useConnection` does)
* change the `WithConnector` and `WithConnectors` component to use the connector transformation function
* enhance the `transformConnectionResponse` to run the new transformation function on the connection's connector, and retrieve the `isTechPreview` flag from the transformed connector object.

kahboom replied:
>Do we really need to accept all these types? Steps don't have a connection, and neither Extensions do.

That was in reference to this feedback: https://github.com/syndesisio/syndesis-react/pull/409#discussion_r291590677
I'll change it to `Connector` 🤷‍♀

riccardo-forina replied:
Yes, that was a step toward a better refactoring. My suggestion aims to not having to rely on those helpers across the app and enhance the model so that the information is already available for the consumer at all times. If you do what I suggested you can enhance the `IUIStep` interface to also have `isTechPreview` and `isConfigRequired` and you can set those properties to false for steps and extensions [here](https://github.com/syndesisio/syndesis-react/blob/02625588f49dd054095fa8a19d809c099e1aa71f/app/ui-react/syndesis/src/modules/integrations/components/editor/utils.ts#L57) where the `IUIStep` object is generated.

kahboom replied:
I'm following but am not sure if I'm doing this properly, please see my changes. Also, was easier to start with `WithConnectors.tsx` first.

riccardo-forina reviewed (changes requested):
I agree with Stan, we either do a full refactoring to smooth all the raw edges, or not.
I'm ok with continuing working on this 'til everything is DRYed up and makes sense.

kahboom committed:
Rm reference to isConfigurationRequired from EditorSteps

kahboom committed:
Rm reference to helpers from EditorSteps

kahboom committed:
Add transformConnectorsResponse and other improvements

kahboom committed:
Add configRequired and isTechPreview props to utils and set to false for Extensions and Steps

kahboom committed:
Make props for IConnectionOverview optional

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2439/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

riccardo-forina commented on app/ui-react/syndesis/src/modules/connections/components/Connections.tsx:
```diff
@@ -13,11 +13,17 @@ import { Translation } from 'react-i18next';
 import { UIContext } from '../../../app';
 import { ApiError, EntityIcon } from '../../../shared';
 
+export interface IConnectionOverview extends ConnectionOverview {
```
Yes something like this, only this should be in the `extra.d.ts` and should be referenced by the `useConnection`, etc. components/hooks to explain their return type.

Oh and no need to specify `board`, it's already part of the `ConnectionOverview` interface.

kahboom replied:
I pushed up some changes, it's a bit difficult to test with staging being down. I moved the interface and the `board` prop, updated the reference in `useConnection` and `Connections.tsx`, but if I missed anything let me know.

riccardo-forina reviewed (commented):
Check the comment up here

kahboom committed:
Move IConnectionOverview to extra.d.ts and rm board prop

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2453/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

kahboom committed:
Add techPreview to useConnector

kahboom committed:
Refactor tech preview check in ConnectorsPage.tsx

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2482/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

kahboom committed:
Refactor the bejesus out of Connection and ConnectionOverview

riccardo-forina commented:
PR Storybook available [here](https://circleci.com/api/v1.1/project/github/syndesisio/syndesis-react/2565/artifacts/0/home/circleci/repo/app/ui-react/doc/index.html)

riccardo-forina commented on app/ui-react/packages/api/src/useConnection.tsx:
```diff
@@ -1,8 +1,16 @@
+export const transformConnectionResponse = (
+  connection: IConnectionOverview
+) => {
   return {
```
I think here we should get the updated connector like `const connector = transformConnectorResponse(connection.connector!);` and use it after.
`isTechPreview` would become `isTechPreview: connector.isTechPreview`

riccardo-forina commented on app/ui-react/syndesis/src/modules/connections/components/Connections.tsx:
```diff
@@ -17,7 +17,7 @@ export interface IConnectionsProps {
-  connections: ConnectionOverview[];
+  connections: IConnectionOverview[];
 
   getConnectionHref(connection: ConnectionOverview): H.LocationDescriptor;
```
can we upgrade this to `IConnectionOverview`?

riccardo-forina commented on app/ui-react/syndesis/src/modules/connections/components/Connections.tsx:
```diff
@@ -17,7 +17,7 @@ export interface IConnectionsProps {
+  connections: IConnectionOverview[];
 
   getConnectionHref(connection: ConnectionOverview): H.LocationDescriptor;
   getConnectionEditHref?(connection: ConnectionOverview): H.LocationDescriptor;
```
can we upgrade this to `IConnectionOverview`?

riccardo-forina commented on app/ui-react/syndesis/src/modules/connections/components/Connections.tsx:
```diff
@@ -77,25 +77,11 @@ export class Connections extends React.Component<IConnectionsProps> {
-                                    configurationRequired={
-                                      configurationRequired
-                                    }
+                                    configurationRequired={c.configRequired}
```
🎉

riccardo-forina commented on app/ui-react/syndesis/src/modules/connections/components/Connections.tsx:
```diff
@@ -135,7 +121,7 @@ export class Connections extends React.Component<IConnectionsProps> {
                                         : undefined
                                     }
-                                    techPreview={isTechPreview}
+                                    techPreview={c.isTechPreview}
```
🥳

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/EditorStepsWithToolbar.tsx:
```diff
@@ -0,0 +1,113 @@
+
+export interface IEditorStepsWithToolbarProps
+  extends IEditorStepsProps,
+    Pick<IConnectionsListViewProps, 'createConnectionButtonStyle'> {
```
I don't think this `Pick` is needed at all

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/interfaces.ts:
```diff
@@ -29,7 +29,9 @@ export enum DataShapeDirection {
 
 export interface IUIStep extends StepKind {
   board?: ConnectionBulletinBoard;
+  configRequired?: boolean;
```
I'd make this required, just to stay on the safe side and highlight any point where we missed passing the updated structure.

riccardo-forina commented on app/ui-react/syndesis/src/modules/integrations/components/editor/interfaces.ts:
```diff
@@ -29,7 +29,9 @@ export enum DataShapeDirection {
   board?: ConnectionBulletinBoard;
+  configRequired?: boolean;
   connector?: Connector;
+  isTechPreview?: boolean;
```
I'd make this required, just to stay on the safe side and highlight any point where we missed passing the updated structure.

riccardo-forina reviewed (changes requested):
Great progress, this is almost done! I left a few comments :P

kahboom committed:
stricter interfaces, upgrade to IConnectionOverview in Connections.tsx, more refactoring in useConnection

kahboom committed:
adapt utils toUIStep to accommodate for isTechPreview and configRequired flags

kahboom committed:
Rename configRequired to isConfigRequired

kahboom committed:
Fix ConnectionCard test
